From e5cab8bf5bbe9fb87f9df5329e50cb2ce218b2e8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20Storsj=C3=B6?= <martin@martin.st>
Date: Sat, 13 May 2023 23:04:22 +0000
Subject: [PATCH 1/3] [lli] Export the MinGW chkstk function from the lli
 executable

This allows all ExecutionEngine tests pass in MinGW build configurations.
---
 llvm/tools/lli/lli.cpp | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/llvm/tools/lli/lli.cpp b/llvm/tools/lli/lli.cpp
index 33b2bf869893..279e6b515fad 100644
--- a/llvm/tools/lli/lli.cpp
+++ b/llvm/tools/lli/lli.cpp
@@ -1178,3 +1178,32 @@ Expected<std::unique_ptr<orc::ExecutorProcessControl>> launchRemote() {
       llvm::orc::SimpleRemoteEPC::Setup(), PipeFD[1][0], PipeFD[0][1]);
 #endif
 }
+
+// For MinGW environments, manually export the __chkstk function from the lli
+// executable.
+//
+// Normally, this function is provided by compiler-rt builtins or libgcc.
+// It is named "_alloca" on i386 (plus the symbol ends up with one extra
+// leading underscore), "___chkstk_ms" on x86_64, and "__chkstk" on arm/aarch64.
+//
+// When Orc tries to resolve symbols at runtime, for MSVC cases it finds the
+// symbol "__chkstk" exported from kernelbase.dll - however the MinGW specific
+// symbol names aren't available that way. Therefore, manually export
+// the relevant symbol from lli, to let it be found at runtime during tests.
+//
+// For real JIT uses, the real compiler support libraries should be linked
+// in, somehow; this is a workaround to let tests pass.
+#ifdef __MINGW32__
+// This is a MinGW version of #pragma comment(linker, "...") that doesn't
+// require compiling with -fms-extensions.
+#if defined(__i386__)
+static __attribute__((section(".drectve"), used)) const char export_chkstk[] =
+    "-export:_alloca";
+#elif defined(__x86_64__)
+static __attribute__((section(".drectve"), used)) const char export_chkstk[] =
+    "-export:___chkstk_ms";
+#else
+static __attribute__((section(".drectve"), used)) const char export_chkstk[] =
+    "-export:__chkstk";
+#endif
+#endif
-- 
2.34.1

